util/graph_test.go

```diff
@@ -0,0 +1,70 @@
+// Copyright 2016 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package util
+
+import (
+	"reflect"
+	"testing"
+)
+
+type testTraversal struct {
+	g       map[int]map[int]struct{}
+	visited map[int]struct{}
+}
+
+func (t *testTraversal) Edges(x T) []T {
+	r := []T{}
+	for v := range t.g[x.(int)] {
+		r = append(r, v)
+	}
+	return r
+}
+
+func (t *testTraversal) Equals(a, b T) bool {
+	return a.(int) == b.(int)
+}
+
+func (t *testTraversal) Visited(x T) bool {
+	_, ok := t.visited[x.(int)]
+	t.visited[x.(int)] = struct{}{}
+	return ok
+}
+
+func TestGraphDFS(t *testing.T) {
+	g := map[int]map[int]struct{}{
+		1: map[int]struct{}{
+			2: struct{}{},
+		},
+		2: map[int]struct{}{
+			3: struct{}{},
+			4: struct{}{},
+		},
+		3: map[int]struct{}{},
+		4: map[int]struct{}{
+			1: struct{}{},
+		},
```

Comment:
Maybe have a graph with edges...  1 -> 2, 3->2.  If we explore 1 then 2, and then 3 we'll encounter 2 again, which is visited, but there are no cycles.